### PR TITLE
perf: avoid unnecessary byte/string conversion

### PR DESCRIPTION
**Pull request by Juneezee**
### Why the changes in this PR are needed?

We can use alternative functions/methods to avoid unnecessary byte/string conversion calls.

### What are the changes in this PR?

1. `(*regexp.Regexp).Match([]byte(<a string>))` -> `(*regexp.Regexp).MatchString(<a string>)` [^1]
2. `utf8.RuneCount([]byte(<a string>))` -> `utf8.RuneCountInString(<a string>)` [^2]:
3. `(*bytes.Buffer).WriteString(string(<a []byte>))` -> `(*bytes.Buffer).Write(<a []byte>)` [^3]

[^1]: https://pkg.go.dev/regexp#Regexp.MatchString
[^2]: https://pkg.go.dev/unicode/utf8#RuneCountInString
[^3]: https://pkg.go.dev/bytes#Buffer.Write

### CODE REVIEW

File: topdown/regex.go [hunk 1]
```diff
@@ -46,7 +46,7 @@ func builtinRegexMatch(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Te
 	if err != nil {
 		return err
 	}
-	return iter(ast.BooleanTerm(re.Match([]byte(s2))))
+	return iter(ast.BooleanTerm(re.MatchString(string(s2))))
```
Comment (ashutosh-narkar) [inline topdown/regex.go:49]: Nit: This one we can keep as-is I think since we're converting anyways.

**Review by srenatus — APPROVED** [review]
Thanks!

**Review by ashutosh-narkar — COMMENTED** [review]
@Juneezee thanks for the contribution. One small comment and we can get this in. Thanks!

File: topdown/regex.go (hunk 1 above)
Comment (Juneezee) [reply topdown/regex.go:49]: Correct if I'm wrong. `s2` is a `ast.String` (which is actually a `string`). So casting `string(s2)` doesn't incur a run-time cost like `[]byte(s2)` does [^1]

https://github.com/open-policy-agent/opa/blob/e903f5d815d1dae1a37ddaaef7deb5628bf5fb50/ast/term.go#L763-L764

image

[^1]: https://stackoverflow.com/a/32253871/7902371

**Review by ashutosh-narkar — APPROVED** [review]

**Pull request merged**